Add support for media proxying (for MSC3910)

Half-Shot opened the pull request:
[MSC3910](https://github.com/matrix-org/matrix-spec-proposals/blob/rav/propsal/content_tokens_for_media/proposals/3910-content-tokens-for-media.md) will introduce the requirement to authenticate media downloads, which will leave bridges like matrix-appservice-irc, which presently rely on *unauthenticated* media to function.

To combat this, this PR introduces a transparent proxy service that will allow unauthenticated media access to continue, albeit with more control over the scope and lifetime of the media. This is done by creating and signing some metadata, that is then provided as a URL for the media. This avoids the need for unbounded storage of media records in the bridge.

Problems:
- [ ] IRC relies upon these IDs being small, whereas right now we're using a lump of JSON which is anything but. We could add a gzip stage to see if that helps the duplication, or use a different format entirely.

Half-Shot committed:
Add support for media proxying

Half-Shot committed:
MSC3911

Half-Shot committed:
Add media proxy tests

Half-Shot committed:
Emit media proxy

tadzik commented:
> IRC relies upon these IDs being small, whereas right now we're using a lump of JSON which is anything but



Addressed this one in https://github.com/matrix-org/matrix-appservice-bridge/pull/498 with a tightly packed data format – brings `mxc://example.com/some_media` token from 208 to 128 bytes, with the same HMAC signature length.

tadzik commented:
https://github.com/tadzik/matrix-appservice-bridge/tree/tadzik/media-proxy brings this up to `develop`, along with integrating https://github.com/matrix-org/matrix-appservice-bridge/pull/498 and a few other minor fixes that makes it usable in the bridges.

Few things still worth figuring out:

1. The `ttl` option is interpreted as the number of miliseconds, even though seconds would be more natural. I'm okay with either, but to avoid confusion I'd put it in the name (`ttl_seconds`, `ttl_msec` or something)
2. In the spirit of the urls being as short as they can be, how about we switch from `/v1/media/download/` to `/v1/md/` or something along those lines? There's no reason for us to track the Matrix Spec URLs in any way here, is it?

tadzik committed:
Binary-encode media proxy URLs rather than base64'ing a JSON (#498)

Saves us 80 bytes for the test data (down from 208 to 128),
while leaving room for future changes.

tadzik committed:
Merge branch 'develop' into hs/media-proxy

tadzik committed:
changelog

tadzik committed:
Break a line to appease the linter

tadzik committed:
Merge branch 'develop' into hs/media-proxy

tadzik commented on src/components/media-proxy.ts:
```diff
@@ -0,0 +1,192 @@
+            next();
+        });
+        this.internalRouter.get('/health', this.getHealth.bind(this));
+        // TODO "/v1/media/thumbnail/xyz"
```
I'm fine with keeping this TODO (or even dropping it entirely), since this is our internal API anyway – we don't need to match the MSC here.

Half-Shot replied:
I'd just drop it and see if we need it in practice.

Half-Shot commented on package.json:
```diff
@@ -60,6 +60,7 @@
     "@types/nopt": "^3.0.32",
     "@types/jsbn": "^1.2.33",
     "@types/pkginfo": "^0.4.3",
+    "@types/uuid": "^8.3.1",
```
weird, are we even using this?

tadzik replied:
Nope, must be a merge fail.

tadzik committed:
Drop a leftover @types/uuid dependency

tadzik committed:
Drop a TODO we don't need
